### PR TITLE
fix: pre-request script variable hostname certificate resolution [INS-4733]

### PR DESCRIPTION
**Pull request by ryan-willis**
This fixes two distinct, but closely related issues:

When using a variable reference in the request URL for the hostname and a pre-request script with non-empty contents, client certs would be filtered using the un-rendered template text before sending the request and yield an error (part 1 of the new smoke test).

When gathering context to send a request that has a pre-request script with non-empty contents, the first cert in the list was always selected, disregarding the disabled flag and the hostname on the cert (part 2 of the new smoke test).

### CODE REVIEW

**Review by marckong — DISMISSED** [review]
lgtm!

**Review by marckong — DISMISSED** [review]

File: packages/insomnia-sdk/src/objects/insomnia.ts [hunk 1]
```diff
@@ -167,16 +168,31 @@ export async function initInsomniaObject(
         localVars: localVariables,
     });
 
-    const existClientCert = rawObj.clientCertificates != null && rawObj.clientCertificates.length > 0;
-    const certificate = existClientCert && rawObj.clientCertificates[0] ?
+    const reqUrl = toUrlObject(rawObj.request.url);
```
Comment (ihexxa) [inline packages/insomnia-sdk/src/objects/insomnia.ts]: This makes sense to me to move this part from below, pls notice that the url parsing is not reliable for all cases as there could be templates in it (there's one card for it). such as representing both host and path with tag `{{_.srvr}}{{ _.path }}`, then this url will not be rendered as expected.

File: packages/insomnia-sdk/src/objects/insomnia.ts [hunk 2]
```diff
@@ -167,16 +168,31 @@ export async function initInsomniaObject(
...
+    const renderedHost = variables.replaceIn(host);
+
+    const renderedBaseUrl = toUrlObject(`${reqUrl.protocol}//${renderedHost}`);
+
+    const filteredCerts = filterClientCertificates(rawObj.clientCertificates || [], renderedBaseUrl.toString());
+    const existingClientCert = filteredCerts != null && filteredCerts.length > 0 && filteredCerts[0];
+    const certificate = existingClientCert ?
         {
-            disabled: rawObj.clientCertificates[0].disabled,
+            disabled: existingClientCert.disabled,
```
Comment (ihexxa) [inline packages/insomnia-sdk/src/objects/insomnia.ts:184]: You might notice that there seems no perfect way to transform it between 2 sides, I'm thinking if we could just leave the certificate undefined at the beginning, if user specified the cert in the script, we prepend to the cert list, or return the original certs, pls let me know what you think.

File: packages/insomnia-sdk/src/objects/insomnia.ts [hunk 3]
```diff
@@ -167,16 +168,31 @@ export async function initInsomniaObject(
...
+            .filter(param => !param.disabled)
+            .map(param => ({ key: param.name, value: param.value }))
+    );
+
+    // hack: sanitize a template reference if it exists in the hostname
+    // so we can filter certificates without a full render on the request
+    const host = reqUrl.getHost().replace(/{{\s*\_\./g, '{{');
+    const renderedHost = variables.replaceIn(host);
+
+    const renderedBaseUrl = toUrlObject(`${reqUrl.protocol}//${renderedHost}`);
```
Comment (ihexxa) [inline packages/insomnia-sdk/src/objects/insomnia.ts]: There could be one edge case here when user doesn't specify the protocol then `reqUrl.protocol` will be empty.

File: packages/insomnia-smoke-test/tests/smoke/mtls.test.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  // ensure disabling the cert actually disables it
+  await page.getByRole('button', { name: 'Add Certificates' }).click();
+  await page.locator('[data-test-id="client-certificate-toggle"]').click();
+  await page.getByRole('button', { name: 'Done' }).click();
+  await page.getByLabel('Request Collection').getByTestId('pet 2').press('Enter');
+
+  await page.getByRole('button', { name: 'Send', exact: true }).click();
+  await expect(statusTag).toContainText('401 Unauthorized');
+  await expect(responseBody).toContainText('Client certificate required');
+
```
Comment (ihexxa) [inline packages/insomnia-smoke-test/tests/smoke/mtls.test.ts:72]: 👍

**Review by ihexxa — COMMENTED** [review]
Added some comments and feel free to let me know your thoughts, also thanks for looking into this complicated issue.

File: packages/insomnia-sdk/src/objects/insomnia.ts (hunk 1 above)
Comment (ryan-willis) [reply packages/insomnia-sdk/src/objects/insomnia.ts]: Good point. I tried to avoid running the url through a full render but it looks like we might need to do that to cover this case

File: packages/insomnia-sdk/src/objects/insomnia.ts (hunk 3 above)
Comment (ryan-willis) [reply packages/insomnia-sdk/src/objects/insomnia.ts]: I see. I'll default to `http`

File: packages/insomnia-sdk/src/objects/insomnia.ts (hunk 1 above)
Comment (ryan-willis) [reply packages/insomnia-sdk/src/objects/insomnia.ts]: Now that I'm looking at this again, I _could_ perform the template render hack prior to calling `getHost`, what do you think?

File: packages/insomnia-sdk/src/objects/insomnia.ts (hunk 2 above)
Comment (ryan-willis) [reply packages/insomnia-sdk/src/objects/insomnia.ts:184]: I'd much rather return original certs, but I'm thinking we revamp this later

**Review by ihexxa — APPROVED** [review]

File: packages/insomnia-sdk/src/objects/insomnia.ts (hunk 1 above)
Comment (ihexxa) [reply packages/insomnia-sdk/src/objects/insomnia.ts]: We also have a plan to improve the `replaceIn` method, let's optimize it later.

File: packages/insomnia-sdk/src/objects/insomnia.ts (hunk 2 above)
Comment (ihexxa) [reply packages/insomnia-sdk/src/objects/insomnia.ts:184]: Yes we can revamp it later and the prepend operation is for those cases who would like to update the cert through script.
